### DISCUSSION TITLE
Using PyInstaller to create an application using ffmpeg (Mac)

### DISCUSSION THREAD

**Discussion by organfreeman36** [pyinstaller/pyinstaller#8089] [PyInstaller]
Hi, I'm currently trying to turn my script into a .app with pyinsatller. It has a customtkinter GUI and several modules. The program uses ffmpeg for some audio file conversion. I've tried bundling ffmpeg (the latest build from the ffmpeg website). However when the program is bundled the resulting .app runs but gets stuck on the file conversion stage that uses ffmpeg. I also can't see ffmpeg in the application bundle.

Here is my spec file:
# -*- mode: python ; coding: utf-8 -*-

block_cipher = None

a = Analysis(['GUI.py'],
             pathex=['/Users/post_5/Desktop/AAC Master/ArchieProject-main'],
             binaries=[('/Users/post_5/Desktop/AAC Master/ArchieProject-main/ffmpeg/ffmpeg', '.')],
             datas=[],
             hiddenimports=[],
             hookspath=[],
             hooksconfig={},
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)

pyz = PYZ(a.pure, cipher=block_cipher)

exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [],
          name='AAC',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=True,
          console=True,
          icon=None,
          runtime_tmpdir=None,
          target_arch=None)

app = BUNDLE(exe,
             name='AAC.app',
             icon=None,
             bundle_identifier='com.aac.aac',
             console=True)

**Comment by bwoodsend** [comment]
What do you mean by stuck? Is it hanging or raising an exception? Can you temporarily put the conversion code in its own script and freeze that so that we can rule out any GUI related issues?

**Reply by organfreeman36** [reply]
Sorry I wasn't very clear. The app launches, you can navigate the GUI and it functions completely normally until you use the app to convert some audio files (which uses ffmpeg) at which point the app hangs on that frame. The program runs correctly as an executable but I assume that's because ffmpeg is installed on the system

**Comment by rokm** [comment]
> I also can't see ffmpeg in the application bundle.

In .app bundle generated by PyInstaller v6, it should be collected into the `AAC.app/Contents/Frameworks` directory (in earlier versions, it would probably be in `Contents/MacOS`). 

At run-time in your code, you should reference it as `os.path.join(sys._MEIPASS, 'ffmpeg')` or `os.path.join(os.path.dirname(__file__), 'ffmpeg')` (the second example is for referencing from the entry-point script).

**Reply by organfreeman36** [reply]
I've checked and I'm using pyinstaller version 6.1.0. The frameworks directory is empty when creating the app

I'm using this code to try and use ffmpeg when the app is frozen:

def get_ffmpeg_path():
    if getattr(sys, 'frozen', False):
        return os.path.join(sys._MEIPASS, 'ffmpeg')
    else:
        return '/opt/homebrew/bin/ffmpeg'

**Reply by rokm** [reply]
> I've checked and I'm using pyinstaller version 6.1.0. The frameworks directory is empty when creating the app

What are the contents of the .app like, then?

**Reply by rokm** [reply]
And the other directories?

**Reply by organfreeman36** [reply]
![Screenshot 2023-11-10 at 09 47 08](https://github.com/pyinstaller/pyinstaller/assets/61791446/d3c2d19d-b7c5-4e2c-be04-1621c4bd740d)

**Reply by organfreeman36** [reply]
![Screenshot 2023-11-10 at 09 48 29](https://github.com/pyinstaller/pyinstaller/assets/61791446/0ac60fa7-e20c-4da1-95d5-ed1b7e3a128a)

**Reply by organfreeman36** [reply]
Sorry, not sure if the other screenshot uploaded. MacOS just contains the AAC executable

**Reply by rokm** [reply]
Did you create a onefile .app bundle instead of an onedir one?

**Reply by organfreeman36** [reply]
I think so? I used '--pyinstaller GUI.spec' to make the application from the above spec file

**Reply by rokm** [reply]
Then that's why you don't see anything except for the executable. You should use onedir instead.

**Reply by organfreeman36** [reply]
I've created the application with onedir. It still has the same issue

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "threading.py", line 1038, in _bootstrap_inner
  File "threading.py", line 975, in run
  File "GUI.py", line 69, in run_conversion
  File "temp_converter.py", line 847, in main
  File "temp_converter.py", line 379, in wav
  File "multiprocessing/pool.py", line 367, in map
  File "multiprocessing/pool.py", line 774, in get
FileNotFoundError: [Errno 2] No such file or directory: '/Users/post_5/Desktop/AAC Master/ArchieProject-main/dist/GUI/_internal/ffmpeg'

**Reply by organfreeman36** [reply]
Sorry I'd forgotten to add ffmpeg to the binaries of the new spec file. It seems to be packaged correctly now. I am getting other errors now when ran as an application however:

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "threading.py", line 1038, in _bootstrap_inner
  File "threading.py", line 975, in run
  File "GUI.py", line 69, in run_conversion
  File "temp_converter.py", line 847, in main
  File "temp_converter.py", line 379, in wav
  File "multiprocessing/pool.py", line 367, in map
  File "multiprocessing/pool.py", line 774, in get
subprocess.CalledProcessError: Command '['ffprobe', '-v', 'error', '-select_streams', 'a:0', '-show_entries', 'stream=channels,sample_rate', '-of', 'default=noprint_wrappers=1:nokey=1', '/Users/post_5/Desktop/AAC/Book Chapters/Output Folder/WAV/004_Part1_Chapter2_Book_Title.wav']' died with <Signals.SIGABRT: 6>.

**Reply by rokm** [reply]
That's the `ffprobe` process dying for whatever reason. If you manually try to run that command using the `ffprobe` collected in the `Contents/Frameworks`, does it work, or does it also die?

**Reply by organfreeman36** [reply]
I think I've fixed the problem, I wasn't referencing ffprobe correctly

I'm a little confused why I have to use --onedir and not --onefile. --onedir does not create a .app bundle, is it not possible to create a .app with the ffmpeg depdencies?

**Reply by rokm** [reply]
`--onefile` does not create an .app bundle, either. It creates a self-extracting POSIX executable, which is not an .app bundle. Proper macOS .app bundles are created by adding `--windowed` to your PyInstaller command. And while .app bundles can have onefile executable, there is no point in that (plus it causes issues with codesigning and notarization for app store).
